# InSituPy: Python Package for the Analysis of Xenium In Situ Data

InSituPy is a Python package designed to facilitate the analysis of in situ transcriptomics data. With InSituPy, you can process, visualize, and analyze spatially resolved transcriptomics data obtained through various in situ techniques, enabling comprehensive exploration of spatial gene expression patterns within tissues.

## Installation

To install InSituPy, follow these steps:

1. Clone the repository to your local machine:

   ```bash
   git clone https://github.com/jwrth/InSituPy.git
   ```

2. Navigate to the cloned repository:

   ```bash
   cd insitupy
   ```

3. Install the package using `pip`:

   ```bash
   pip install .
   ```

Make sure you have Python 3.6 or later installed on your system.

## Getting Started

For detailed instructions on using InSituPy, refer to the [official documentation](https://InSituPy.readthedocs.io). The documentation provides comprehensive guides on installation, usage, and advanced features.

## Tutorials

Explore our tutorials to learn how to use InSituPy effectively:

1. [Basic Data Processing](https://InSituPy.readthedocs.io/en/latest/tutorials/basic_processing.html) - Learn how to preprocess raw data and prepare it for analysis.
2. [Spatial Visualization](https://InSituPy.readthedocs.io/en/latest/tutorials/spatial_visualization.html) - Visualize spatial gene expression patterns using interactive plots and heatmaps.
3. [Advanced Analysis](https://InSituPy.readthedocs.io/en/latest/tutorials/advanced_analysis.html) - Dive into advanced analysis techniques, including spatial clustering and cell type identification.

## Features

- **Data Preprocessing:** InSituPy provides functions for normalizing, filtering, and transforming raw in situ transcriptomics data.

- **Spatial Visualization:** Explore gene expression patterns in spatial context using interactive visualizations and heatmaps.

- **Spatial Clustering:** Identify spatially distinct clusters of cells or regions within tissues.

- **Cell Type Annotation:** Automatically annotate cell types based on spatial gene expression profiles.

- **Integration with Single-Cell Data:** Integrate in situ transcriptomics data with single-cell RNA-seq data for comprehensive analysis.

- **Interactive Plots:** Create interactive plots to explore spatial gene expression patterns across different samples and conditions.

## Contributing

Contributions are welcome! If you find any issues or have suggestions for new features, please open an [issue](https://github.com/jwrth/InSituPy/issues) or submit a pull request.

## License

InSituPy is licensed under the [MIT License](LICENSE).

---

**InSituPy** is developed and maintained by the InSituPy team. We appreciate your feedback and hope this package enhances your in situ transcriptomics analysis experience!
